src/Microsoft.Health.Fhir.Core/Features/Persistence/Orchestration/BundleOrchestratorOperation.cs

```diff
@@ -132,26 +133,55 @@ public async Task<UpsertOutcome> AppendResourceAsync(ResourceWrapperOperation re
                 // Await for the merge async task to complete merging all resources.
                 var ingestedResources = await _mergeAsyncTask;
 
-                if (ingestedResources.TryGetValue(identifier, out DataStoreOperationOutcome dataStoreOperationOutcome))
+                DataStoreOperationOutcome dataStoreOperationOutcome;
+
+                // Attempt 1: Retrieve from the list of merged records, the one with the same identifier from current thread.
+                if (!ingestedResources.TryGetValue(identifier, out dataStoreOperationOutcome))
                 {
-                    if (!dataStoreOperationOutcome.IsOperationSuccessful)
+                    // Attemp 2: Edge case scenario:
+                    // Under racing conditions, it's possible that the same record (<resourceType>/<id>) is present in two bundles running at the same time.
+                    // One record is updated first than the second, increasing the version of the record, and then updated once more by the second bundle.
+                    // As the version is higher than the expected, the local identifier in the current thread does not match, and an alternative search for
+                    // the combination <resourceType>/<id> is required.
+                    var ingestedResourcesById = ingestedResources.Where(i => i.Key.ResourceType == identifier.ResourceType && i.Key.Id == identifier.Id).ToList();
+
+                    int countOfResourcesFound = ingestedResourcesById.Count;
+                    if (countOfResourcesFound == 1)
                     {
-                        throw dataStoreOperationOutcome.Exception;
+                        // Edge Case scenario: resource was updated by another bundle concurrently, but it was also updated by the current bundle, increasing its version.
```

Comment:
Where are we **"increasing the version"**? or you mean **"updating the version"**?

Reply:
Can we add a log to know the resource was sent twice and we have already updated it, just for our records or debugging? ofc if you think it's helpful

Reply:
The version is increased on SQL Server side. 

I can add a log to this edge case scenario mentioning we handled it.